File: finder4j-backtrack/src/test/java/com/github/k0zka/finder4j/backtrack/termination/FirstSolutionTerminationStrategyTest.java
package com.github.k0zka.finder4j.backtrack.termination;

import org.junit.Assert;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.runners.MockitoJUnitRunner;

import com.github.k0zka.finder4j.backtrack.State;
import com.github.k0zka.finder4j.backtrack.Step;

@RunWith(MockitoJUnitRunner.class)
public class FirstSolutionTerminationStrategyTest {

	@Mock
	State state;

	@Test
	public void onSolution() {
		FirstSolutionTerminationStrategy<State, Step<State>> strategy = new FirstSolutionTerminationStrategy<State, Step<State>>();
		Assert.assertNull(strategy.getSolution());
		Assert.assertFalse(strategy.stop(state));
		strategy.onSolution(state);
		Assert.assertEquals(state, strategy.getSolution());
		Assert.assertTrue(strategy.stop(state));
	}
}

File: finder4j-backtrack/src/test/kotlin/com/github/k0zka/finder4j/backtrack/termination/FirstSolutionTerminationStrategyTest.kt
package com.github.k0zka.finder4j.backtrack.termination

import com.github.k0zka.finder4j.backtrack.State
import com.github.k0zka.finder4j.backtrack.Step
import com.nhaarman.mockito_kotlin.mock
import org.junit.Assert
import org.junit.Test

class FirstSolutionTerminationStrategyTest {

	internal var state: State = mock()

	@Test
	fun onSolution() {
		val strategy = FirstSolutionTerminationStrategy<State>()
		Assert.assertNull(strategy.solution)
		Assert.assertFalse(strategy.stop(state))
		strategy.onSolution(state)
		Assert.assertEquals(state, strategy.solution)
		Assert.assertTrue(strategy.stop(state))
	}
}
